cloudinary-test-common/src/main/java/com/cloudinary/test/AbstractStructuredMetadataTest.java

```diff
@@ -0,0 +1,241 @@
+
+        String message = null;
+        ApiResponse res = null;
+        try {
+            // should fail
+            dateMetadataField.setDefaultValue(illegalValue);
+            res = api.addMetadataField(dateMetadataField);
+            // this line should not be reached if all is working well, but when it's not we still want to clean it up:
+            metadataFieldExternalIds.add(res.get("external_id").toString());
+        } catch (BadRequest e) {
+            message = e.getMessage();
+        }
+
+        assertEquals(message, "default_value is invalid");
+
+        // should work:
+        dateMetadataField.setDefaultValue(legalValue);
+        res = api.addMetadataField(dateMetadataField);
+        metadataFieldExternalIds.add(res.get("external_id").toString());
+    }
+
+    @Test
+    public void testListFields() throws Exception {
+        StringMetadataField stringField = newFieldInstance("testListFields");
+        addFieldToAccount(stringField);
+
+        ApiResponse result = cloudinary.api().listMetadataFields();
+        assertNotNull(result);
+    }
+
+    @Test
+    public void testGetMetadata() throws Exception {
+        ApiResponse fieldResult = addFieldToAccount(newFieldInstance("testGetMetadata"));
+        ApiResponse result = api.metadataFieldByFieldId(fieldResult.get("external_id").toString());
+        assertNotNull(result);
+        assertEquals(fieldResult.get("label"), result.get("label"));
+    }
+
+    @Test
+    public void testUpdateField() throws Exception {
+        ApiResponse fieldResult = addFieldToAccount(newFieldInstance("testUpdateField"));
+        assertNotEquals("new_def", fieldResult.get("default_value"));
+        StringMetadataField field = new StringMetadataField();
+        field.setDefaultValue("new_def");
+        ApiResponse result = api.updateMetadataField(fieldResult.get("external_id").toString(), field);
+        assertNotNull(result);
+        assertEquals("new_def", result.get("default_value"));
+    }
+
+    @Test
+    public void testDeleteField() throws Exception {
+        ApiResponse fieldResult = addFieldToAccount(newFieldInstance("testDeleteField"));
+        ApiResponse result = api.deleteMetadataField(fieldResult.get("external_id").toString());
+        assertNotNull(result);
+        assertEquals("ok", result.get("message"));
+    }
+
+    @Test
+    public void testUpdateDatasource() throws Exception {
+        SetMetadataField setField = createSetField("testUpdateDatasource");
+        ApiResponse fieldResult = addFieldToAccount(setField);
+        MetadataDataSource.Entry newEntry = new MetadataDataSource.Entry("id1", "new1");
+        ApiResponse result = api.updateMetadataFieldDatasource(fieldResult.get("external_id").toString(), Collections.singletonList(newEntry));
+        assertNotNull(result);
+        assertEquals("new1", ((Map) ((List) result.get("values")).get(0)).get("value"));
```

Comment:
Check what happens when adding an entry that already exists

Reply:
Again, your call but I was trying to avoid making this a full backend integration/regression test. There isn't even a 'correct' behaviour in the spec to test against...